TLM/TLM/Manager/Impl/LaneConnectionManager.cs

```diff
@@ -80,39 +101,108 @@ public class LaneConnectionManager
 
-        public bool HasConnections(uint sourceLaneId, bool startNode) {
+        /// <summary>
+        /// Determines if the given lane has outgoing connections
+        /// </summary>
+        /// <param name="sourceLaneId"></param>
+        /// <returns></returns>
+        public bool HasOutgoingConnections(uint sourceLaneId, bool startNode) {
             if (!Options.laneConnectorEnabled) {
                 return false;
             }
 
-            int nodeArrayIndex = startNode ? 0 : 1;
+            if(!ValidateLane(sourceLaneId))
+                return false;
 
-            return Flags.laneConnections[sourceLaneId] != null &&
-                   Flags.laneConnections[sourceLaneId][nodeArrayIndex] != null;
+            var targets = connections_.GetConnections(sourceLaneId, startNode);
+            int n = targets?.Length ?? 0;
+            for(int i = 0; i < n; ++i) {
+                uint targetLaneId = targets[i].LaneId;
+                if(ValidateLane(targetLaneId))
+                    return true;
+
+            }
+            return false;
+        }
+
+        public bool HasConnections(uint laneId, bool startNode) => HasConnections(laneId, laneId.ToLane().GetNodeId(startNode));
+
+        /// <summary>
+        /// Determines if the given lane has incoming/outgoing connections
+        /// </summary>
+        public bool HasConnections(uint laneId, ushort nodeId) {
+            bool startNode = laneId.ToLane().IsStartNode(nodeId);
+            return HasOutgoingConnections(laneId, startNode) || HasIncommingConnections(laneId, nodeId);
+        }
+
+        public bool HasIncommingConnections(uint targetLaneId, ushort nodeId) {
+            if(!ValidateLane(targetLaneId))
+                return false;
+
+            ref NetNode node = ref nodeId.ToNode();
+            for(int segmentIndex = 0; segmentIndex < 8; ++segmentIndex) {
+                ushort segmentId = node.GetSegment(segmentIndex);
+                if(segmentId == 0)
+                    continue;
+
+                ref NetSegment segment = ref segmentId.ToSegment();
+                if(!segment.IsValid()) {
+                    HandleInvalidSegmentImpl(segmentId);
+                    // even if segment is invalid we should still remove invalid lanes so we go through.
+                }
+
+                bool startNode = segment.IsStartnode(nodeId);
+                var laneInfos = segment.Info.m_lanes;
+                uint laneId = segment.m_lanes;
+                for(int laneIndex = 0; laneIndex < laneInfos.Length && laneId != 0; ++laneIndex) {
+                    ref NetLane lane = ref laneId.ToLane();
+                    if(laneId != targetLaneId) {
+                        var laneInfo = laneInfos[laneIndex];
+                        if(laneInfo.m_laneType.IsFlagSet(LANE_TYPES) &&
+                            laneInfo.m_vehicleType.IsFlagSet(VEHICLE_TYPES) &&
+                            connections_.IsConnectedTo(laneId, targetLaneId, startNode)) {
```

Comment:
looks like hidden nested loop :/

Reply:
yep. This is what happens when I remove the backward connection. I have to search for it :).